global/scripts/fsl_version.shlib

```diff
@@ -10,20 +16,170 @@
+
+    if [ -f ${fsl_version_file} ]
+    then
+        fsl_version=`cat ${fsl_version_file}`
+        log_Msg "INFO: Determined that the FSL version in use is ${fsl_version}"
+    else
+        log_Msg "ERROR: Cannot tell which version of FSL you are using."
+        exit 1
+    fi
+
+    eval $__functionResultVar="'${fsl_version}'"
+}
+
+#
+# Function Description
+#  Standardize the FSL-version number (in argument 1) for easy
+#  version check and comparison
+#  
+#  This function take a version number (e.g. 6.0.4)
+#  and format it as Num.Num.Num.Num (e.g. 6.0.4.0). 
+#  This make the version comparison/check easier 
+#  (e.g. when comparing 6.0.4.0 with 6.0.4.1). 
+#  Here we assume that FSL-version is not using more than
+#  4 digits and 3 dots (M.X.Y.Z), otherwise increase max_ndots. 
+#
+#  Some formatting examples:
+#  6 => 6.0.0.0 
+#  6.0 => 6.0.0.0
+#  6.0.4 => 6.0.4.0
+#  6.0.4.12 => 6.0.4.12 
+#
+# Usage Example:
+#  fsl_ver2=$(fsl_version_format "6.0.4")
+#
+fsl_version_format(){
+    local fv="${1}"
+    local ndots
+    # maximum number of dots in FSL-version string
+    local max_ndots=$MAX_NDOTS_IN_FSLVERSION
+
+    ndots=$(awk -F"." '{print NF-1}' <<< "$fv")
+
+    while [ "$ndots" -lt "$max_ndots" ]; do
+        fv="$fv".0
+        ndots=$(awk -F"." '{print NF-1}' <<< "$fv")
+    done
+    
+    echo "$fv"
+}
+
+# Function Description
+#  Check if the current FSL version (the FSL version in use) 
+#  is at least equal to the version given in argument 1 ($1). 
+#
+#  This function echoes string "true" (if FSL-version >= argument1) 
+#  or "false" (if FSL-version < argument1).
+#  It is up to the user to decide what to do with this information. 
+# 
+# Usage Example:
+#  ret_val=$(fsl_minimum_required_version "6.0.7.2")
+#  if [ "$ret_val" == "false" ]; then
+#      log_Msg "ERROR: FSL minimum required version is 6.0.7.2. 
+#      exit 1
+#  else 
+#      log_Msg "INFO: FSL version > 6.0.7.2"
+#      echo "do something..."
+#  fi 
+# 
+fsl_minimum_required_version(){
+    local min_ver="${1}" 
+    local cur_ver 
+    local min_ver_array
+    local cur_ver_array
+    local ndots_cur_ver
+    local ndots_min_ver
+    # maximum number of dots in FSL-version string
+    local max_ndots
+
+    # get current FSL version
+    fsl_version_get cur_ver >> /dev/null
+
+    # if the 2 version strings are identical, no need to test further
+    if [ "$cur_ver" == "$min_ver" ]; 
+    then
+        echo "true"
+        return 0
+    fi
+
+    # split version by dots
+    IFS=. read -a cur_ver_array <<< "$cur_ver"
+    IFS=. read -a min_ver_array <<< "$min_ver"
+
+    # count the dots
+    ndots_cur_ver=$(awk -F"." '{print NF-1}' <<< "$cur_ver")
+    ndots_min_ver=$(awk -F"." '{print NF-1}' <<< "$min_ver")
+
+    # get the maximum number of dots
+    if [ "$ndots_cur_ver" -ge "$ndots_min_ver" ]; 
+    then
+        max_ndots=$ndots_cur_ver
+    else 
+        max_ndots=$ndots_min_ver
+    fi
+
+    for k in $(seq 0 $max_ndots)
```

Comment:
You should be able to stop at the minimum number of dots instead of the maximum, if you don't treat `3.0.0` as being equal to `3.0`.

Reply:
dpkg implements "x.0 is newer than x", and while it has operators that are described as "treat empty version as later", that doesn't change these cases, they literally mean "if the whole version string is the empty string, it is greater than any nonempty string":
```
$ if dpkg --compare-versions 3.0 'lt-nl' 3.0.0; then echo yes; fi
yes
```

I suggest this, which doesn't need to add zeros:
```bash
min_fields=$((${#cur_ver_array[@]}))
if ((${#req_ver_array[@]} < min_fields))
then
    min_fields=$((${#req_ver_array[@]}))
fi
for ((i = 0; i < min_fields; ++i))
    if ((cur_ver_array[i] < req_ver_array[i])); then echo false; return; fi
    if ((cur_ver_array[i] > req_ver_array[i])); then echo true; return; fi
end
if ((${#cur_ver_array[@]} >= ${#req_ver_array[@]})); then echo true; return; fi
echo false; return
```

Reply:
Right, I have treated "x.0 as equal to x" but I am also fine with "x.0 is newer than x". 
I'll try and rerun test.

Reply:
The comparison is now "x.0 is newer than x".
So before modification it was returning
```
Test:       6.0.7.6.0 <= 6.0.7.6 is true
Test:     6.0.7.6.0.0 <= 6.0.7.6 is true
```
and now these cases are false: 
```
Test:       6.0.7.6.0 <= 6.0.7.6 is false
Test:     6.0.7.6.0.0 <= 6.0.7.6 is false
``` 
All other cases returns the same results. 
I attached test results and and test script. 

[fsl_version_test_before.txt](https://github.com/Washington-University/HCPpipelines/files/14319071/fsl_version_test_before.txt)
[fsl_version_test_after.txt](https://github.com/Washington-University/HCPpipelines/files/14319072/fsl_version_test_after.txt)
[test_modification_fslversion_functions.sh.txt](https://github.com/Washington-University/HCPpipelines/files/14319076/test_modification_fslversion_functions.sh.txt)